pytorch_toolkit/formula_recognition/im2latex/models/text_recognition_heads/attention_based.py

```diff
@@ -0,0 +1,301 @@
+"""
+MIT License
+
+Copyright (c) 2019 luopeixiang
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+
+ Copyright (c) 2020 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+
+import math
+from collections import namedtuple, OrderedDict
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn import init
+
+from ...data.vocab import END_TOKEN, START_TOKEN
+
+INIT = 1e-2
+Candidate = namedtuple('candidate', 'score, dec_state_h, dec_state_c, output, targets, logits')
+
+
+class TextRecognitionHead(nn.Module):
+    def __init__(self, out_size, configuration):
```

Comment:
Yes, you are right, configuration is a dict. Passing arguments as a dict is more convenient (except of `out_size`, because it is a vocab size, and it is calculated after loading vocab file). Moreover, configuration as a dict is also passed into other classed (backbone, model, etc), so this is somehow unified way to pass parameters for class.

Reply:
> Passing arguments as a dict is more convenient

This is a design choice and I doubt its advantages. What benefits do you see?

Reply:
At least in this way I do not have to unpack dictionary outside the class constructor. If I do so, I would have duplicate code in any place I call the class constructor. If this is head class, it is only one place (in `Im2latexModel` constructor), but if I change this behavior and unpack parameters before every call of class constructor, I would have at least for duplicate pieces of code (in train, eval, demo & export scripts) whenever I call the model constructor. If I change the behavior only in head constructor, it would look strange that head constructor interface is different from other model parts interfaces.

Reply:
What do you mean saying "unpack dictionary"? Could you provide a sample?

Reply:
Say, I describe head config in the config file. If I understand you correctly, you want to make several positional arguments in the head constructor:
```
class TextRecognitionHead(nn.Module):
    def __init__(self, out_size, encoder_hidden_size, decoder_hidden_size, beam_width):
```
This leads to the fact that we would have to call the head constructor in the following way in the training script
```
head_config = config.get('head')
out_size = head_config.get("out_size")
encoder_hidden_size = head_config.get("encoder_hidden_size")
decoder_hidden_size = head_config.get("decoder_hidden_size")
beam_width = head_config.get("beam_width")
model = Im2latexModel(backbone_type, backbone, out_size, head_config)
```
Also, if I would like to add some new parameter in the future, I would have to change behavior in the `Head` class and in the `Im2latexModel` class, which could lead to a potential bug due to duplication of the code

Reply:
> you want to make several positional arguments in the head constructor

Not positional, but named ones:

```python
class TextRecognitionHead(nn.Module):
    def __init__(self, out_size, encoder_hidden_size=111, decoder_hidden_size=222, beam_width=333):
        # ...
        pass
```

And then

```python
head = TextRecognitionHead(out_size, **config.get('head', {}))
```

Reply:
Oh, I did not think of this way, thank you

Reply:
Done

Reply:
How about `ResNetLikeBackbone` as well? Is it possible to refactor it alike?

Reply:
Yes, done